main/src/library/Array.flix

```diff
@@ -57,10 +57,1390 @@ namespace Array {
     ///
     /// Helper for toList.
     ///
-    def toListHelper(a: Array[a], i: Int): List[a] & Impure =
+    def toListHelper(a: Array[a], i: Int32): List[a] & Impure =
         if (i == length(a))
             Nil
         else
             a[i] :: toListHelper(a, i + 1)
 
-}
+    ///
+    /// Returns `Some(x)` if `x` is the first element of `a`.
+    ///
+    /// Returns `None` if `a` is empty.
+    ///
+    pub def head(a: Array[a]): Option[a] & Pure =
```

Comment:
Perhaps surprising, any operation that touches an array (except getting its length) must be marked as impure.

The reason is that we define purity as "equal under substitution". If we have the program:

```
let a = [1, 2, 3];
a[0] = 42;
Array.head(a)
```

then it is *not* equivalent to:

```
let a = [1, 2, 3];
a[0] = 42;
Array.head([1, 2, 3])
```

Thus a function like `head`, perhaps unfortunately, but with good reason, must be marked as impure.

Interestingly, getting the length of an array is pure. This is because the length of an array, once allocated, cannot be changed.

Reply:
This comment with purity applies to other functions as well, but a quick search for `& Pure` or similar should let you find them easily.

Reply:
Bonus comment: In future, we might have some light-weight ownership system that would allow us to actually mark a function such as `head` as pure provided that no-one else can mutate `a`.